### ISSUE TITLE
Scaling Tacotron model for different data sizes 

### ISSUE THREAD

**Issue by AliNGatGeeks** [espnet/espnet#4288]
Hi Guys 
I am trying to train the Tacotron model on a dataset much smaller than Ljspeech dataset. about 2400 samples. 
The model overfits after a couple of epochs using the config file used for training TTS on Ljspeech dataset. 
Using default training config, the model doesn't overfit using default config until 37th or 38th epoch but the results using either Griffin lim algorithm or Parallel WaveGAN is not good at all. The Parallel waveGAN produced very good results after about 100K steps . 

Any pointers that could help to change the config file for this dataset?(or anything that could help) 
dataset: Javanese single female dataset

**Comment by kan-bayashi** [comment]
First, please check these points:
https://github.com/espnet/espnet/tree/master/egs2/TEMPLATE/tts1#why-the-model-cannot-be-trained-well-with-my-own-dataset

Several comments:
- Most of the cases, all you need is cleaning.
- In the case of TTS, validation best model does not produce the best results.  The overfitted model is often better than the validation best model in terms of the quality.
- Did you check the attention weight? If it is diagonal, the model cannot generate reasonable results even if the loss itself is decreased.

**Comment by AliNGatGeeks** [comment]
Thank you for tips. I started by working on text and cleaning it 
I started a new training. I have three folder showing attention images 


![training](https://user-images.githubusercontent.com/96764993/165077561-8651ebd8-4d94-4245-a13f-be44c9be9838.png)

this a sample from each folder. Do these seam good? I cant interpret the figures my self... any references to read more about them? or a few words would be much appreciated

**Comment by kan-bayashi** [comment]
It looks good except for the last part of the first example.
(Maybe it includes silence at the end of audio)
You can see bad and good examples in https://github.com/espnet/espnet/issues/1937

**Comment by AliNGatGeeks** [comment]
Thank you for you comment. 
I continued working my case but I get bad results. 
\
Task I am doing: train German TTS model with Tacotron 2 and ParallelWaveGAN v1 
\
I think both models are working well individually, but when I combine them I get very poor result. 

Generated wav:
['Im kompletten Wohnzimmer war der Fußboden mit Teppich ausgelegt.'](https://drive.google.com/file/d/1J30yeJHySoDLcQDZ_2tKLyNxPDx7I6lO/view?usp=sharing)
['Den Namen habe ich schon einmal gehört'](https://drive.google.com/file/d/1C3-aT8U7-sJ8zWVfuN7JhQoObGkrFGAd/view?usp=sharing)
\
I would really appreciate it, If you can help me figure out what I am missing or doing wrong 
I have provided detail of what I have done below:

**Training Tacotron 2 for Mel2Text:** 
<ul>
<li>I used a dataset of 22k samples for TTS  (German) </li>
<li>For mel2text I used ESPNET2 recipe </li>
<li>model tacotron2 </li>
<li> I followed the recipe from step 2 to 6</li>
<li>token type : character encoding, </li>
<li>200 epochs of training and only changed batch bins to fit into memory. </li>
</ul>
<br>

<p   style="text-align:center;">some results  on test set.</p>

![image](https://user-images.githubusercontent.com/96764993/166099170-006fa896-de24-4638-8736-4c99415fa3dd.png)
<hr>

**Parallel WaveGAN v1**
\
Then I used the same dataset to train a Parallel WaveGAN using this repo. 

<ul>
<li>model : Parallel wavegan v1</li>
<li> for training I copied the dataset into the template folder   </li>
<li>22K samples and --num_dev 250 --num_eval 400</li>
<li> didnt change anything in the config file </li>
<li>After 75000 steps results seem reasonable </li>
</ul>

link to some generated files : 

[generated wave1](https://drive.google.com/file/d/1LKlTqTP3F92fPHqH5P_3Egdqq4FmDxLn/view?usp=sharing)
[generated wave 2](https://drive.google.com/file/d/1a2TLRZuFlkki7T4gMR5RxajPeZrsxrJ8/view?usp=sharing) 

<hr>

**Test**

Following the recipe I used 
`
ttsv = Text2Speech.from_pretrained(model_file=model_file,vocoder_file=vocoder_file) 

text = 'Im kompletten Wohnzimmer war der Fußboden mit Teppich ausgelegt.'
wav = ttsv(text)["wav"]
sr = 22050
wav1 = wav.cpu().detach().numpy()

import IPython.display as ipd
ipd.Audio(wav1, rate=sr) # load a NumPy array
`

Generated wav: 
['Im kompletten Wohnzimmer war der Fußboden mit Teppich ausgelegt.'](https://drive.google.com/file/d/1J30yeJHySoDLcQDZ_2tKLyNxPDx7I6lO/view?usp=sharing)
['Den Namen habe ich schon einmal gehört'](https://drive.google.com/file/d/1C3-aT8U7-sJ8zWVfuN7JhQoObGkrFGAd/view?usp=sharing)

**Comment by kan-bayashi** [comment]
Please check the following points:
- Make sure the sample without the vocoder (i.e., griffin lim) sounds reasonable 
- Make sure you used the totally same feature extraction setting (sampling rate / hop size / window size / min max ...)
- 75000 step is still in the pretraining phase, PWG will improve until 400k steps.

**Comment by AliNGatGeeks** [comment]
Hi 
thank you for your comments. 
I checked and I had made a bad mistake! my sampling rate was 16K for Tacotron and 22K for PWGAN! So I have to train again ... 

however, currently I have Tacotron model trained with16K sampling rate, do you recognize the following examples decoded with Griffin lim as reasonable? 

<h3>Griffin lim example sound reasonable?</h3>

[Griffin lim: Das ist doch viel witziger.wav](https://drive.google.com/file/d/1Nd5fyJHaIzcEnu6rTaFEf2lXgKXfdOF7/view?usp=sharing)
[Griffin lim:  werde nicht wieder lachen.wav](https://drive.google.com/file/d/1UXAAnVOc3iOrquLJ8h0TsF4vGBhEUVRM/view?usp=sharing)
[Griffin lim: Afghanistan gilt kein generelles Tempolimit auf Autobahnen.wav](https://drive.google.com/file/d/11Q-pP1MKuqAB5n0iogLLv8iO-lQYSl3D/view?usp=sharing)

[Original: Das ist doch viel witziger.wav](https://drive.google.com/file/d/1vKIqjo3rjt4YvDv54q8ChdkBBwcw8Yxz/view?usp=sharing)
[Original: ich werde nicht wieder lachen.wav](https://drive.google.com/file/d/1bU6GIHDhaaqIMMghMQDPv7cgDRdcyq3j/view?usp=sharing)
[Original: In Afghanistan gilt kein generelles Tempolimit auf Autobahnen.wav](https://drive.google.com/file/d/1atwgQYU6rV761NjfkifKQrEBKkBseHe5/view?usp=sharing)

<hr>

<h3>Matching the settings </h3>
These are the setting I am currently using to train again. Tacotron (Left) and PWGAN (Right)... have I got it right?(anything I have to check ?)
<br>
![ask about settings](https://user-images.githubusercontent.com/96764993/167302980-2777945b-b65b-47d1-8260-63f0e1d1f988.png)
<hr>


<hr>
<h2>Tokenization </h2>
I am training on  German language but using character tokenization... I havent been able to install phonemizer yet... 

Can I manually change these? to add special German characters or remove some thing from it?
like: 
Ä/ä, Ö/ö, Ü/ü) and one is a consonant ligature of ⟨s⟩ and ⟨z⟩ (ß; called Eszett (sz) or scharfes S, "sharp s").
 


![tokensGermant](https://user-images.githubusercontent.com/96764993/167303187-904d2d86-697a-4a30-92b1-f4b83730a725.png)

**Comment by kan-bayashi** [comment]
> Griffin lim as reasonable?

I could not understand the lang but it sounds reasonable.

>  Tacotron (Left) and PWGAN (Right)

The feature extraction setting seems fine.

> Can I manually change these? to add special German characters or remove some thing from it? like: Ä/ä, Ö/ö, Ü/ü) and one is a consonant ligature of ⟨s⟩ and ⟨z⟩ (ß; called Eszett (sz) or scharfes S, "sharp s").

You can edit `dump/token_list/hogehoge/tokens.txt`, which is created in stage 4.
If you change it, you need to run the training again.
(And if you run the stage 4, it will be overwritten, be careful.)

> do we also need to remove silence and split samples for training the decoder?

You mean text 2 mel part? Yes, you should.
For the vocoder, the effect is not so serious.

**Comment by AliNGatGeeks** [comment]
Hi 

- I trained my  Tacotron2 model again. 
- I removed silence from the wav files at begging ,end and in between signal 
- Also changed the sampling rate to 22050 so it matches  with the vocoder. 
- I used previously trained vocoder (not trimmed and split wavs)
- However, The combined output of the Tacotron2 and Vocoder (Parallel WGAN) is **very noisy!** 


I have provided some details below, and would really appreciate any help or comments.  


Some examples: 

Sentence: In Afghanistan gilt kein generelles Tempolimit auf Autobahnen
[Original wav](https://drive.google.com/file/d/1atwgQYU6rV761NjfkifKQrEBKkBseHe5/view?usp=sharing)
[decoded with Griffin Lim](https://drive.google.com/file/d/1CFjqI2P22nFKLtJ9LfKC4cv5mgaz-QAp/view?usp=sharing)
[decoded using Parallel WaveGAN](https://drive.google.com/file/d/1pIexLT4xiJ-45AS-2Vkdh2U3q2Ju4Lhh/view?usp=sharing)

Sentence: 'Die Nutzung dieses Angebotes ist für Sie kostenlos.'
[Original wav](https://drive.google.com/file/d/1nuKXm2Ytjn11T9OvSxBWLo0njWa9chRr/view?usp=sharing)
[decoded with Griffin Lim](https://drive.google.com/file/d/1jMtkAVNtX8vGfoFXle5IVPfzuNNPGMh2/view?usp=sharing)
[decoded with Parallel WaveGan](https://drive.google.com/file/d/1w8ZF0gZNqyewuz5PvYr_KGIeNz-TZIzD/view?usp=sharing)


<h2>Predictions of Parallel WaveGAN during training</h2>
<p>The vocoder is also trained up to 400K steps and here are some of the prediction produced during training: 
to me the output seem good!?!(do they?)</p>
Sample1 : 

[Original wav]( https://drive.google.com/file/d/11qQJGMZFzAPPiDYhPNzlaNZJF5kTAR2n/view?usp=sharing)
[Generated wave](https://drive.google.com/file/d/1z-o6M-IbhLVHNBLNvFokQGEXl-1ZO6yx/view?usp=sharing)


Sample 2: 
[Original wav](https://drive.google.com/file/d/1bF8JDa1NbWJE1208jtaGcDhMmgJHn4l8/view?usp=sharing)
[Generated wave](https://drive.google.com/file/d/1slHg-maaO_5c8OmgVE1gor-2HrCvRDu1/view?usp=sharing)

<h2>Configuration Files </h2>

<h3>Tacotron </h3>

[Train_tacotron2.yml](https://drive.google.com/file/d/1O_VBGFKN54_b3r9WENFbAd-1ekLGxqfp/view?usp=sharing)
[Config generated by training step.yml](https://drive.google.com/file/d/1-533Jq-B7okKC58-Z77yVWiqokPS-nW9/view?usp=sharing)

<h3>Vocoder train </h3>

[parallel_wavegan.v1.yaml](https://drive.google.com/file/d/1O1RBRNeeS-pUFfD6-bhWbzzVFomrQBFX/view?usp=sharing)

[Config Generated by training step,yml](https://drive.google.com/file/d/1O1RBRNeeS-pUFfD6-bhWbzzVFomrQBFX/view?usp=sharing)

![PWGANsampl1image](https://user-images.githubusercontent.com/96764993/168230475-9fec1fc7-37b8-4112-a651-bb7188ac4c96.png)

**Comment by kan-bayashi** [comment]
How did you perform the decoding with the trained vocoder?
Please write the command as is.

**Comment by AliNGatGeeks** [comment]
`
 #load models 
import os
model_root = '/home/ec2-user/SageMaker/espnet/egs2/German_Male/tts1/exp/tts_train_tacotron2_raw_char_tacotron'

model = '115epoch.pth'
model_file = os.path.join(model_root,model)
print(model_file)

vocoder= 'checkpoint-400000steps.pkl'
root = '/home/ec2-user/SageMaker/espnet/egs2/German_Male/tts1/decoder/'
vocoder_file = os.path.join(root,vocoder)
print(vocoder_file)



text = 'Die Nutzung dieses Angebotes ist für Sie kostenlos.'

ttsv = Text2Speech.from_pretrained(model_file=model_file,vocoder_file=vocoder_file)


wav = ttsv(text)["wav"]
# sr = 22050
wav1 = wav.cpu().detach().numpy()


plt.figure(figsize=(14, 5))
librosa.display.waveshow(wav1, sr=22050)

ipd.Audio(wav, rate=22050) # load a NumPy array

(and then I download the file from Jupyter)
`

**Comment by kan-bayashi** [comment]
`'/home/ec2-user/SageMaker/espnet/egs2/German_Male/tts1/decoder/'` include `stats.h5`?

**Comment by AliNGatGeeks** [comment]
**WOW**

Indeed it wasn't!!!!!!!!!!!!! I added this file and now its working !!!! 
Thank you very very much
